src/wiringx.c

```diff
@@ -105,6 +88,99 @@ static struct spi_t spi[2] = {
 	} while (0)
 #endif
 
+#ifndef __FreeBSD__
+/* SPI Bus Parameters */
+struct spi_t {
+	uint8_t mode;
+	uint8_t bits_per_word;
+	uint16_t delay;
+	uint32_t speed;
+	int fd;
+} spi_t;
+
+/* SPI states list implementation */
+struct spi_id_t {
+	uint8_t device;
+	uint8_t channel;
+};
+
+struct spi_list_t {
+	struct spi_id_t id;
+	struct spi_t data;
+};
+
+static struct spi_list_t *spilist = NULL;
+static int nrspilist = 0;
+
+static int spi_new(struct spi_id_t id) {
+	int entry = -1;
+	void *tmp = NULL;
+	int i = 0;
+
+	/* loop over existing states to detect potential duplicates and reusable entries */
+	for(i = 0; i < nrspilist; i++) {
+		/* detect dupplicate */
+		if(memcmp(&spilist[i].id, &id, sizeof(struct spi_id_t)) == 0) {
```

Comment:
Are you sure this is correct. The `id` is a local variable in the that only lives in the `wiringXSPISetup` stack as long as the function lasts. So as soon as we reach the end of this function, the memory reserved on the stack is freed. Therefor, we can never reliably check if the id is similar to another id, because the memory area of the first id instance is already undefined.

So, in order to get this to work, you need to explicitly allocate memory for the id in the `wiringXSPISetup` and make sure it get's properly freed when it's usage has become obsolete.

I also think it's better not to rely on the struct contents, but explicitly compare the device and channel properties.

Reply:
oh we don't really check for duplicates. We want to know if spilist already has an entry with the same (by value) id. Because that means all fields of the id struct need to be identical, memcmp is safe.
One id struct lives in wiringXSPISetup, initialized as struct spi_id_t id = { device, channel };
The other instances are each non-pointer members of spi_list_t, and are assigned by vaklue, not pointer: spilist[entry].id = id;

Reply:
> The other instances are each non-pointer members of spi_list_t, and are assigned by vaklue, not pointer: spilist[entry].id = id;

Does copy a struct from one instance to another work like this?

Reply:
Yes. It is an assignment with the same type.
However there is a point to be made about comparing fields rather than memcmp.